api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,629 @@
+	}
+
+	// these ports are automatically used by the ecs agent
+	defaultPorts := []int{
+		22,
+		2376,
+		2375,
+		51678,
+		51679,
+	}
+
+	resource := &models.ResourceProvider{}
+	resource.ID = "<new instance>"
+	resource.InUse = false
+	resource.UsedPorts = defaultPorts
+	resource.AvailableMemory = instanceType
+	return resource, nil
+}
+
+func (e *EnvironmentScaler) getContainerResourceFromDeploy(deployID string) ([]models.ResourceConsumer, error) {
+	// use some kind of deploy cache
+	if consumers, ok := e.deployCache[deployID]; ok {
+		return consumers, nil
+	}
+
+	input := &ecs.DescribeTaskDefinitionInput{}
+	input.SetTaskDefinition(deployID)
+	output, err := e.Client.ECS.DescribeTaskDefinition(input)
+	if err != nil {
+		return nil, err
+	}
+
+	consumers := make([]models.ResourceConsumer, len(output.TaskDefinition.ContainerDefinitions))
+	for i, d := range output.TaskDefinition.ContainerDefinitions {
+		var memory bytesize.Bytesize
+
+		if aws.Int64Value(d.MemoryReservation) != 0 {
+			memory = bytesize.MiB * bytesize.Bytesize(aws.Int64Value(d.MemoryReservation))
+		}
+
+		if aws.Int64Value(d.Memory) != 0 {
+			memory = bytesize.MiB * bytesize.Bytesize(aws.Int64Value(d.Memory))
+		}
+
+		ports := []int{}
+		for _, p := range d.PortMappings {
+			if aws.Int64Value(p.HostPort) != 0 {
+				ports = append(ports, int(aws.Int64Value(p.HostPort)))
+			}
+		}
+
+		consumers[i] = models.ResourceConsumer{
+			ID:     "",
+			Memory: fmt.Sprintf("%v", memory), //todo: translate bytesize.ByteSize to string correctly
+			Ports:  ports,
+		}
+	}
+
+	e.deployCache[deployID] = consumers
+	return consumers, nil
+}
+
+func (e *EnvironmentScaler) scaleTo(environmentID string, scale int, unusedProviders []*models.ResourceProvider) (int, error) {
+	clusterName := addLayer0Prefix(e.Config.Instance(), environmentID)
+
+	input := &autoscaling.DescribeAutoScalingGroupsInput{}
+	input.SetAutoScalingGroupNames([]*string{&clusterName})
+
+	asg, err := e.Client.AutoScaling.DescribeAutoScalingGroups(input)
+	if err != nil {
+		return 0, err
+	}
+
+	currentCapacity := int(aws.Int64Value(asg.AutoScalingGroups[0].DesiredCapacity))
+
+	switch {
+	case scale > currentCapacity:
+		log.Printf("Environment %s is attempting to scale up to size %d", clusterName, scale)
+		return e.scaleUp(clusterName, scale, asg.AutoScalingGroups[0])
+	case scale < currentCapacity:
+		log.Printf("Environment %s is attempting to scale down to size %d", clusterName, scale)
+		return e.scaleDown(clusterName, scale, asg.AutoScalingGroups[0], unusedProviders)
+	default:
+		log.Printf("Environment %s is at desired scale of %d. No scaling action required.", clusterName, scale)
+		return currentCapacity, nil
+	}
+}
+
+func (e *EnvironmentScaler) scaleUp(ecsEnvironmentID string, scale int, asg *autoscaling.Group) (int, error) {
+	maxCapacity := int(aws.Int64Value(asg.MaxSize))
+	if scale > maxCapacity {
+		input := &autoscaling.UpdateAutoScalingGroupInput{}
```

Comment:
[nit] For consistency, since you have `log.Printf()` statements for `scaleDown` when scale < minCapacity, I think similar logging statements should be present in this method when scale > maxCapacity.